_posts/2020-12-25-manual-gates-git-flow-in-a-wig.md

```diff
@@ -0,0 +1,145 @@
+
+We've made no mystery of our dislike of GitFlow and our preference for Trunk-Based Development (Myerscough, T. 2019). We want to improve the flow of software into production thereby allowing users to get value faster. The way to do this responsibly is to automate as much testing and verification as possible and reduce the number of steps that involve human intervention. This is easiest to obtain when working in the Trunk-Based model. Merging from branch-to-branch-to-branch introduces additional steps that we feel does not add any value to the software development process, as everything should land on master eventually anyway.
+
+But it isn't just us. The creator of GitFlow has somewhat denounced it recently (Driessen, V. 2010), explaining that it was designed for a world in which consumers typically hosted versions of the software themselves. This is very different from the present where the typical delivery model is via SaaS-based web and mobile applications. The CEO of GitLab has raised similar concerns (Sijbrandij, S. 2015), and GitHub themselves publish their own methodology which is more in line with Trunk-Based Development (GitHub. 2020). DevOps pioneer, author of the Phoenix Project, and fellow of DevOps Research and Assessment (DORA), Jez Humble, has had similar misgivings about the GitFlow (Humble, J. 2020).
+
+> never use git flow. It’s a horrible way to solve the problem the creator used it for, and also a really horrible way to build software generally - Jez Humble
+
+Which brings me to the point made in the headline - Manual Gates are functionally GitFlow. If you are attempting to do Trunk-Based-Development, but you have manual gates between stages in your CI/CD pipeline, this is fundamentally the same as GitFlow with merging between environments. In each case someone has to manually inspect the currently deployed environment, decide whether it is healthy, then take some action to promote the software into the next environment. In each case they both have the same impacts on the flow of software into the production environment - waiting for a human to make a decision.
+
+# Are Manual Gates That Bad?
+
+Yes. Sort of. To be honest, manual gates are only bad because 99% of the time they are used to facilitate manual testing. What about the other 1%? I believe the only time a gate is useful is to control when software hits production. You may only want deploy during off-peak times or on certain days for stability reasons e.g. if the software has all the feature it needs for an important event that will happen on Wednesday, it perhaps stands to reason to wait till Thursday before rolling out a non-essential feature.
```

Comment:
The point you've not covered here, is that manual gates encourage batching.  And batching *increases* the risk of failures.

So rather than reducing risk, manual gates are just moving the risk elsewhere - from the risk of small changes causing failures/regressions in production to large batches of changes causing big problems.  Often a result of "somebody else's problem" responsibilities within the software delivery team.

Reply:
Your case for manual gates is also weak (possibly intentional).  It doesn'tacknowledge the fact that the development team should have visibility of such events, and control over the deployments.  Just don't merge code before an important event - you bring the pain forward and make it visible. If you want to keep merging code, then you should be able to deploy anytime.

Reply:
Yeah they should - the point here with the gate is that everything should have already made it through pre-prod testing. The point I'm trying to make is pushing it through the pipeline is all about preparing to go to prod. The gate in this instance is less about the batching risk - the rest of the pipeline should be rock solid and the risk of going to prod in this case should be minimal.

If you don't deploy though - your risk of breaking the service from a deployment drops from minimal to zero. That said - I don't think developers should have to break their normal workflow (merging as soon as possible) for some external factor. e.g. development cadence is decoupled from the release cadence.

I mean, we agree that everything in the head of master should be in a release-able state, this is just charging the line a bit.

Does that make sense?

Reply:
> The gate in this instance is less about the batching risk - the rest of the pipeline should be rock solid and the risk of going to prod in this case should be minimal.
> If you don't deploy though - your risk of breaking the service from a deployment drops from minimal to zero. 
Your risk today may be zero, but you're storing up the technical debt/deployment debt each time you batch.  You have lots of small changes you have confidence in.  But not going all the way to prod means you now have a disconnect.

This is the difference between Continuous Delivery and Continuous Deployment - Continuous Delivery is about release of value to users (always potentially releasable).  Continuous Deployment is about separating deployment from release - to tackle batching risks.

> The point I'm trying to make is pushing it through the pipeline is all about preparing to go to prod.
My view is the point of the pipeline is about getting to prod - your value stream.  Therefore that's only fulfilled once stuff is actually running in production (still a separate concern from released to users).

I think we're both in agreement here though.

>(That being I said i think it's worth perhaps addressing additional batching risk here, and will add something in)
:+1: